File: android/chat-module/src/main/java/expo/modules/vibechatnative/ChatTransport.kt
package expo.modules.vibechatnative

internal data class ChatTransportEvent(
  val topic: String,
  val event: String,
  val payload: Map<String, Any?>,
  val ref: String? = null,
  val joinRef: String? = null,
)

internal interface ChatTransportCallbacks {
  fun onOpen()
  fun onClosed(code: Int, reason: String?)
  fun onError(error: String)
  fun onEvent(frame: ChatTransportEvent)
}

internal interface ChatRealtimeTransport {
  fun connect()
  fun disconnect()
  fun join(topic: String, payload: Map<String, Any?> = emptyMap()): String
  fun leave(topic: String): String
  fun push(topic: String, event: String, payload: Map<String, Any?> = emptyMap()): String
}

File: ios/ChatModule/ChatTransport.swift
import Foundation

@available(iOS 13.0, *)
struct ChatTransportFrame {
  let joinRef: String?
  let ref: String?
  let topic: String
  let event: String
  let payload: [String: Any]
}

@available(iOS 13.0, *)
struct ChatTransportCallbacks {
  let onOpen: () -> Void
  let onClose: (_ code: Int, _ reason: String?) -> Void
  let onError: (_ error: String) -> Void
  let onEvent: (_ frame: ChatTransportFrame) -> Void
}

@available(iOS 13.0, *)
protocol ChatRealtimeTransport: AnyObject {
  func connect()
  func disconnect()
  @discardableResult
  func join(topic: String, payload: [String: Any]) -> String
  @discardableResult
  func leave(topic: String) -> String
  @discardableResult
  func push(topic: String, event: String, payload: [String: Any]) -> String
}
